Refactor our Related Orders Data Store classes to support WC CRUD methods

mattallan opened the pull request:
Fixes #145

## Description

This PR updates our `WCS_Related_Order_Store_Cached_CPT` and `WCS_Related_Order_Store_CPT` classes to use WC CRUD methods to be able to support both WP Posts and High-Performance Order Storage (HPOS) environments.

> **Note** - Doesn't include renaming our classes
>
> This PR doesn't rename our classes to remove the reference to Custom Post Types with the `_CPT` suffix. Renaming these classes and deprecating the class isn't a critical part to this refactor, so I will create a separate issue to tackle this at a later time.

> **Note** - Can't test with HPOS
>
> This PR does not update our `WCS_Post_Meta_Cache_Manager` class which triggers a lot of the functionality in these related orders store classes, this means you cannot test this PR with HPOS enabled. You must be using WP Posts.

### What are these caches?

Subscription related orders (renewals, switch and resubscribe orders) record their relationship in order meta. Historically finding these subscription related orders was costly as it required querying the database for orders with specific meta (a meta query and post join). To fix this in WC Subscriptions 2.3 we introduced a persistent caching layer. In subscription meta we now store a single key to keep track of the subscriptions related orders. 

<img width="548" alt="Screen Shot 2022-11-09 at 4 46 27 pm" src="https://user-images.githubusercontent.com/8490476/200758656-4b3f762e-a571-4203-aadc-a2c28d0e8aeb.png">

Now to get a subscription's list of renewal orders we simply just refer to this single meta key on the subscription rather than multiple values stored on every order. 

These cache classes act as gatekeepers by providing APIs for fetching related orders `get_related_order_ids()` and they also house all the hooks for keeping the cache up to date. If a renewal order is trashed, this class is responsible for removing that record for the subscription's related order cache. 

There are 2 public facing pages that include some details of these classes: 

- [**Subscription Caches** → **_Subscription Related Order Cache_**](https://woocommerce.com/document/subscriptions/develop/cache/#section-1)
- [**What’s New in Subscriptions 2.3** → **_Performance_**](https://woocommerce.com/document/version-2-3/#section-7)

### What's new?

There's a fair bit changing in this PR so let's go through it:
 - New Subscriptions data store functions:
   - `delete_all_metadata_by_key` - Previously we used WP's `delete_metadata()` to remove every cache meta we had stored. There's no equivalent in the WC HPOS data stores so we created our own.
   - ~`get_metadata_by_key` - When fetching related order cache meta data we need to make sure it's fetched directly from the database. This function pulls the metadata directly from the new orders table or if using wp posts, it calls `get_post_meta()`~
 - Replaced instanced of  `get_post_meta`, `update_post_meta`  & `delete_post_meta` with relevant WC CRUD method.
 - Replaced uses of `get_posts()` with new `wcs_get_orders_with_meta_query()`
 - ~New helper function `get_subscription_with_cache_meta()` to help us get an instance of WC Subscriptions with all of our related order cache meta in memory. This solves a problem I was facing with trying to fetch the cache returning an empty array and trying to update the cache, creating a duplicate postmeta entry.~

## How to test this PR

Testing this PR should hopefully be pretty straightforward and the main thing we need to test is making sure the subscription's related-order cache is properly updated when its list of related orders changes, and when an order is updated we also need to make sure the relevant related order meta is updated on the orders.

Flows to test:
1. After processing a renewal order, the renewal order is visible in the related orders table on the Edit Subscription page
2. After processing a renewal order, the new renewal order ID is inside the `_subscription_renewal_order_ids_cache` on the subscription
3. After processing a renewal order, make sure the renewal order has the `_subscription_renewal` and proper subscription ID saved in postmeta.
4. After trashing a renewal order, the renewal order ID is removed from the `_subscription_renewal_order_ids_cache` array in subscription postmeta
5. After trashing a renewal order, the renewal order should not show up in the related orders table on the subscription

## Product impact
<!-- What products will this PR ship in? -->

- [ ] Added changelog entry (or does not apply)
- [x] Will this PR affect WooCommerce Subscriptions? yes/no/tbc, add issue ref
- [x] Will this PR affect WooCommerce Payments? yes/no/tbc, add issue ref
- [x] <!-- 🚨 Deprecations 🚨 --> Added deprecated functions, hooks or classes to the [spreadsheet](https://docs.google.com/spreadsheets/d/1xw9xszcPMnWsp4C8OKZMsLzZob7tOmWT7qMqmEIq314/edit#gid=0)

Commit:
Clean up docblock in WCS_Related_Order_Store_CPT

Commit:
Remove pre WC 3.0 code in WCS_Related_Order_Store_CPT

Commit:
Remove deprecated filter handling from new class introduced in WCS 2.3

The new Related Order data store doesn't need the deprecated filter handling that was deprecated in WC Subscriptions 2.3.2.

Commit:
Update get_related_order_ids() to replace get_posts() with wc_get_orders()

Commit:
Fix docblock comments in WCS_Related_Order_Store_Cached_CPT

Commit:
Update get/update/delete_post_meta() uses to use WC CRUD methods

Commit:
Update delete_caches_for_all_subscriptions() to work for any data store

Previous function was using WP delete_metadata() to delete all rows that have a cache meta key.
This commit creates a new helper in our data store classes to delete this meta data
depending on whether HPOS is enabled

Commit:
replace direct DB queries in related orders datastore/cache classes

Commit:
Merge branch 'trunk' into issue/145-related-orders-data-store

Commit:
Adjust wcs_get_orders query to use known args

Commit:
Read cache meta directly from database with new data store function

Commit:
Remove the ignore props filter so that we properly read all meta before update

Becuase these props are being ignored, when we read a subscription into memory, this memory isn't being stored.
Then when we go to update meta data and save the subscription, we end up with duplicate meta data for cache meta.

Commit:
Replace wcs_get_objects_property() uses with proper getter

Commit:
Always pass around the subscription object instead of subscription ID

Commit:
Introduce and use new function to get an instance of a subscription with cache meta

Commit:
Rename delete_all_meta_data_with_key() to delete_all_metadata_by_key()

mattallan commented:
Tests are currently failing due to `wcs_get_orders_with_meta_query()` not existing

haszari commented:
> This PR updates our `WCS_Related_Order_Store_Cached_CPT` and `WCS_Related_Order_Store_CPT` classes to use WC CRUD methods to be able to support both WP Posts and High-Performance Order Storage (HPOS) environments.

To make this PR easier to review, and to spread expertise about subscriptions code base, can we add some info about the purpose of these classes? I.e. what "related orders" means to subscriptions users. 

@mattallan I'm thinking the PR description could define up front what related orders are in simple terms (i.e. suitable for a naive merchant), but we might also want to flesh out the code docs for the classes as well. If this is already documented feel free to link me to it! Also FYI @james-allan  and @brucealdridge  since Matt is AFK right now.

aprea commented on includes/data-stores/class-wcs-subscription-data-store-cpt.php:
```diff
@@ -549,4 +549,27 @@ public function get_subscription_customer_ids() {
+	 * @param string $meta_key The meta key to delete.
+	 */
+	public function delete_all_metadata_by_key( $meta_key ) {
+		// Set variables to workaround ambiguous parameters of delete_metadata()
```
**Nitpick:** This comment looks like copypasta from the original code as we don't appear to be setting variables here.

james-allan replied:
Fixed in 74cbf070.

aprea commented on includes/data-stores/class-wcs-orders-table-subscription-data-store.php:
```diff
@@ -603,4 +603,59 @@ public function get_subscription_customer_ids() {
+		global $wpdb;
+		$table_name = self::get_meta_table_name();
+
+		$wpdb->query(
```
**Nitpick:** Could we use [`$wpdb->delete()` ](https://developer.wordpress.org/reference/classes/wpdb/delete/)here instead? It probably doesn't make much difference but it seems like an appropriate use case.

james-allan replied:
Yep thats a good idea and it looks a lot cleaner. Fixed in 8f5914b9

james-allan commented:
> To make this PR easier to review, and to spread expertise about subscriptions code base, can we add some info about the purpose of these classes? I.e. what "related orders" means to subscriptions users.

I'll try to flesh out the PR description with some additional details. 

> we might also want to flesh out the code docs for the classes as well. If this is already documented feel free to link me to it!

This is the only documentation ([Subscription Caches → Subscription Related Order Cache](https://woocommerce.com/document/subscriptions/develop/cache/#section-1)) that I think we have. It's on a page about cached data, isn't super in depth and includes details we'd want to remove or rephrase. Specifically the `get_post_meta` reference. The only other documentation I found on these classes comes from the 2.3 release post. [What’s New in Subscriptions 2.3 → Performance](https://woocommerce.com/document/version-2-3/#section-7)

haszari commented:
> I'll try to flesh out the PR description with some additional details.

Thanks @james-allan ! Can you capture that info in docblocks or something more persistent (than PR) while you're there? 🙌

aprea committed:
Merge branch 'trunk' into issue/145-related-orders-data-store

james-allan commented on includes/data-stores/class-wcs-related-order-store-cached-cpt.php:
```diff
@@ -298,8 +324,9 @@ public function delete_caches_for_all_subscriptions( $relation_types = array() )
-	 * @param array $props_to_ignore A mapping of meta keys => prop names.
-	 * @param WCS_Subscription_Data_Store_CPT $data_store
+	 * @param array                           $props_to_ignore A mapping of meta keys => prop names.
+	 * @param WCS_Subscription_Data_Store_CPT $data_store      Subscriptions Data Store
```
Do you know if this is still needed? If a store is using HPOS tables and this doesn't receive a CPT datastore object, I'm assuming we still want to ignore these keys.

james-allan committed:
Update function block and inline comments

james-allan committed:
Use strict conditions when determining if a cache contains an order

I've confirmed locally that the related order caches contain integers and the order ID is an int to so this
being a strict compare is fine.

james-allan committed:
Define variables to resolve ambiguous parameters of delete_metadata()

james-allan committed:
Replace wpdb query with wpdb delete

aprea commented on includes/data-stores/class-wcs-related-order-store-cached-cpt.php:
```diff
@@ -73,165 +75,173 @@ protected function init() {
 
 	/**
-	 * Helper function for setting related order cache.
+	 * Setts a subscription's related order cache for a given relationship.
```
```suggestion
	 * Sets a subscription's related order cache for a given relationship.
```

aprea commented on includes/data-stores/class-wcs-related-order-store-cached-cpt.php:
```diff
@@ -73,165 +75,173 @@ protected function init() {
 
-		$related_order_ids = $this->get_related_order_ids( $subscription, $relation_type );
+		$related_order_ids   = $this->get_related_order_ids( $subscription, $relation_type );
+		$related_order_index = array_search( $order_id, $related_order_ids, true );
```
I just wanted to check something. I can see the strict param was added here (i.e. 3rd param in `array_search`).

Are we certain that both `$order_id` and the elements in `$related_order_ids` will be the same type? i.e. both strings or both integers?

The function docblock says that `$order_id` will be an int. Are we certain the elements in the `$related_order_ids` array will be ints?

james-allan replied:
Yeah I confirmed this locally and included a note in the commit message. https://github.com/Automattic/woocommerce-subscriptions-core/pull/256/commits/55fb511b551e873622e8bd4ac758981d7ad310b3

It might be worth intval'ing each ID just in case.

james-allan committed:
Partial fixes for failing unit tests

aprea commented on includes/data-stores/class-wcs-orders-table-subscription-data-store.php:
```diff
@@ -603,4 +603,53 @@ public function get_subscription_customer_ids() {
+	 *
+	 * @return mixed The meta value.
+	 */
+	public function get_metadata_by_key( $subscription_id, $meta_key, $single = true ) {
```
I'm surprised that Woo core doesn't provide a helper/abstraction for this.

james-allan replied:
They have methods for fetching meta data by key from the order or subscription in memory but because of a quirk with this class (the meta keys are ignored when being read from the database) we needed a way to fetch the data via the key directly from the database. 

In the end I was able to find a way around it in [372af8c](https://github.com/Automattic/woocommerce-subscriptions-core/pull/256/commits/372af8c0d310f0655341f6f551903d2bfeaf5e36) and deleted this function in [7bfa7e2](https://github.com/Automattic/woocommerce-subscriptions-core/pull/256/commits/7bfa7e2b787e3debefbfa31f0691a9c36774ab4c)

haszari replied:
Nice! If the Woo apis should be improved maybe we can log that request with woo core.

james-allan committed:
Update includes/data-stores/class-wcs-related-order-store-cached-cpt.php

Co-authored-by: Chris Aprea <1527164+aprea@users.noreply.github.com>

james-allan committed:
Rework the way we fetch subscription meta data with the related order cache key present

james-allan committed:
Remove the get_metadata_by_key functions as they aren't needed anymore

james-allan committed:
Merge branch 'trunk' into issue/145-related-orders-data-store

james-allan committed:
Merge branch 'trunk' into issue/145-related-orders-data-store

james-allan reviewed (commented):
I've finally been able to get unit tests to pass. 😅  

This class is actually covered pretty extensively with unit test so getting them passing has given me more confidence in these changes. 

I've submitted a bunch of changes to this PR so I'm going to do another round of CR to make sure things are looking in order.

james-allan committed:
Add block comment to explain override_ignored_props

james-allan committed:
Merge branch 'issue/145-related-orders-data-store' of github.com:automattic/woocommerce-subscriptions-core into issue/145-related-orders-data-store

james-allan committed:
More comment changes

james-allan committed:
Pass around a subscription object where possible to avoid getting a new instance every time

james-allan committed:
Case releated order IDs to int before strict comparison

james-allan committed:
Remove double $

james-allan commented:
I noticed the following deprecated hooks have been removed as part of this PR. I've added them to the [spreadsheet](https://docs.google.com/spreadsheets/d/1xw9xszcPMnWsp4C8OKZMsLzZob7tOmWT7qMqmEIq314/edit#gid=0). 


- `wcs_subscriptions_for_renewal_order`
- `wcs_subscriptions_for_switch_order`
- `wcs_subscriptions_for_resubscribe_order`

james-allan reviewed (approved):
I've done some more testing of this on my local site and reviewed the changes again. I'm happy to approve this but you may want to give it a once over too given the extent of my changes while you're afk.

Commit:
Add changelog entries

Commit:
Merge branch 'trunk' into issue/145-related-orders-data-store

mattallan commented:
Thanks @james-allan for all of your adjustments, fixes and feedback on this PR while I was AFK!!

> I noticed the following deprecated hooks have been removed as part of this PR

Thanks for spotting this and adding them to the spreadsheet!

I've added some changelog entries for this PR and performed more testing on the latest changes (see below). I'll wait for GH checks to finish before merging

---

 - [x] Processed a new renewal order.
   - [x] Confirmed the `_subscription_renewal_order_ids_cache` meta on the subscription was updated to include the new renewal order
   - [x] Confirmed the new renewal has the correct subscription ID stored in `_subscription_renewal`
   - [x] Confirmed related orders tables on both the subscription and order are correct.
 - [x] Trashed an existing renewal order.
   - [x] Confirmed the order ID was removed from the `_subscription_renewal_order_ids_cache` meta on the subscription
   - [x] Confirmed related orders tables on both the subscription and order are correct.
 - [x] Resubscribed to a cancelled subscription.
   - [x] Confirmed `_subscription_resubscribe_order_ids_cache` has the correct meta
